File: base/platform/android/src/main/java/com/lynx/base/memory/ProcessMemoryInfo.java
// Copyright 2026 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

package com.lynx.base.memory;

import android.os.Build;
import android.os.Debug;
import android.util.Log;
import com.lynx.base.CalledByNative;
import java.util.Map;

public final class ProcessMemoryInfo {
  private static final String TAG = "ProcessMemoryInfo";

  private ProcessMemoryInfo() {}

  @CalledByNative
  private static String[] getMemoryStats() {
    try {
      Debug.MemoryInfo memoryInfo = new Debug.MemoryInfo();
      Debug.getMemoryInfo(memoryInfo);

      if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
        return new String[] {"summary.total-pss", Integer.toString(memoryInfo.getTotalPss())};
      }

      Map<String, String> stats = memoryInfo.getMemoryStats();
      String[] memoryStats = new String[stats.size() * 2];
      int index = 0;
      for (Map.Entry<String, String> entry : stats.entrySet()) {
        memoryStats[index++] = entry.getKey();
        memoryStats[index++] = entry.getValue();
      }
      return memoryStats;
    } catch (Exception e) {
      Log.w(TAG, "Failed to get memory stats.", e);
      return new String[0];
    }
  }
}

File: base/src/platform/android/process_memory_info.cc
// Copyright 2026 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

#include "base/include/platform/android/process_memory_info.h"

#include "base/include/platform/android/jni_convert_helper.h"
#include "base/include/platform/android/jni_utils.h"
#include "base/platform/android/src/main/jni/gen/ProcessMemoryInfo_jni.h"
#include "base/platform/android/src/main/jni/gen/ProcessMemoryInfo_register_jni.h"

namespace lynxbase {
namespace jni {
bool RegisterJNIForProcessMemoryInfo(JNIEnv* env) {
  return RegisterNativesImpl(env);
}
}  // namespace jni
}  // namespace lynxbase

namespace lynx {
namespace base {
namespace android {

std::vector<std::string> GetProcessMemoryInfo() {
  JNIEnv* env = AttachCurrentThread();
  auto memory_stats = Java_ProcessMemoryInfo_getMemoryStats(env);
  return JNIConvertHelper::ConvertJavaStringArrayToStringVector(
      env, memory_stats.Get());
}

}  // namespace android
}  // namespace base
}  // namespace lynx
